### PR TITLE
[AIRFLOW-1235] Fix webserver's odd behaviour

### PR DESCRIPTION
**Pull request by sekikn**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1235

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

In some cases, the gunicorn master shuts down
but the webserver monitor process doesn't.
This PR add timeout functionality to shutdown
all related processes in such cases.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

tests.core:CliTests.test_cli_webserver_shutdown_when_gunicorn_master_is_killed

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by sekikn** [comment]
CI failed on test_trigger_dag_for_date, which is reported as AIRFLOW-1245. It's irrelevant to this fix since CI failed only on Python 2.7 x PostgreSQL backend.

**Comment by icorbett** [comment]
AIRFLOW-1245 has been merged: https://github.com/apache/incubator-airflow/pull/2325

**Comment by sekikn** [comment]
Thanks for the information @icorbett. I've just rebased the PR.

**Comment by sekikn** [comment]
Rebased on master.

**Comment by Fokko** [comment]
Why isn't this test executed on Travis?

**Comment by sekikn** [comment]
Thanks for the review @fokko, I have some reasons to disable it on Travis:

- AFAIR accessing pidfile doesn't work as expected on Travis for some environmental reason. Strangely, when I tried to run this test on Travis, it seemed to cause another test failure.
- This test uses another thread to kill gunicorn with some delay. It's non-deterministic and can occur transient test failure.

Instead of enabling this on Travis, I ran it locally and confirmed that it succeeded.

**Comment by Fokko** [comment]
@sekikn Thanks for elaboration on the problem. I think we should find a way of isolating these tests. For me, a test that isn't automated, is as valuable as no test. I hope you understand, and this is something at the core of Airflow, so we have to make sure that it is tested well.

**Comment by sekikn** [comment]
I understand your concern. I'll update the PR if I come up with some good solution. Thanks :)

**Comment by Fokko** [comment]
@sekikn Thanks!

**Comment by sekikn** [comment]
Updated PR. Instead of killing gunicorn actually using pidfile, it uses mock.patch to emulate the situation that gunicorn doesn't respond. @Fokko would you take a look?

**Comment by sekikn** [comment]
Updated PR. Removed unnecessary code in newly added test as #3123 was merged.

**Comment by Fokko** [comment]
@sekikn Could you rebase on master?

**Comment by sekikn** [comment]
Rebased on master. CI failure occurred on test_authorized and it's irrelevant to this PR. It's continued for at least 3 days (e.g. https://travis-ci.org/apache/incubator-airflow/builds/351858876).

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -618,17 +623,18 @@ def restart_workers(gunicorn_master_proc, num_workers_expected):
...
+    def wait_until_true(fn, timeout=0):
         """
         Sleeps until fn is true
         """
+        t = time.time()
         while not fn():
+            if 0 < timeout and timeout <= time.time() - t:
+                raise AirflowWebServerTimeout(
+                    "No response from gunicorn master within {0} seconds"
+                    .format(master_timeout))
```
Comment (artwr) [inline airflow/bin/cli.py]: Should this be called just `timeout`?

**Review by artwr — COMMENTED** [review]
Aside from one variable which seems to need renaming, this looks good. Let's fix this and merge. @sekikn

File: airflow/bin/cli.py (hunk 1 above)
Comment (sekikn) [reply airflow/bin/cli.py]: Exactly. I'll update the PR. Thanks!

**Comment by johnarnold** [comment]
+1 I like it.

**Comment by johnarnold** [comment]
@artwr merge?

**Comment by artwr** [comment]
Done

**Comment by sekikn** [comment]
@artwr @johnarnold Thanks a lot!

**Pull request closed without merging**